src/mlpack/methods/reinforcement_learning/environment/maze.hpp

```diff
@@ -0,0 +1,406 @@
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+
+/**
+ * Implementation of Maze task.
+ */
+class Maze
+{
+ public:
+  /**
+   * Implementation of the state of Maze.
+   * Each State is a tuple {(row, column)} in n * n maze matrix
+   */
+  class State
+  {
+   public:
+    /**
+     * Construct a state instance.
+     */
+    State() : data(dimension)
+    { /* Nothing to do here. */ }
+
+    /**
+     * Construct a state instance from given data.
+     *
+     * @param data Data 
+     */
+    State(const arma::vec& data) : data(data)
+    { /* Nothing to do here */ }
+
+    //! Modify the internal representation of the state.
+    arma::vec& Data() { return data; }
+
+    //! Get the internal representation of the state.
+    arma::vec Data() const { return data; }
+
+    //! Encode the state to a column vector.
+    const arma::vec& Encode() const { return data; }
+
+    //! Get value of row index of the node.
+    double Row() const { return data[0]; }
+    //! Modify value of row index of the node.
+    double& Row() { return data[0]; }
+
+    //! Get value of column index of the node.
+    double Column() const { return data[1]; }
+    //! Modify value of column index of the node.
+    double& Column() { return data[1]; }
+
+    //! Dimension of the encoded state.
+    static constexpr size_t dimension = 2;
+
+   private:
+    //! Locally-stored {(row, column)}.
+    arma::vec data;
+  };
+
+  /**
+   * Implementation of action of Maze.
+   */
+  class Action
+  {
+   public:
+   enum actions
+    {
+      left,
+      right,
+      down,
+      up,
+    };
+    
+    // To store the action ( direction to move )
+    Action::actions action;
+
+    // Track the size of the action space.
+    static const size_t size = 4;
+  };
+
+  /**
+   * Construct a maze instance using the given constants.
+   * @param rows number of rows in maze.
+   * @param columns number of columns in maze.
+   * @param maxSteps The number of steps after which the episode
+   *    terminates. If the value is 0, there is no limit.
+   */
+  Maze(const size_t rows = 4,
+        const size_t columns = 4,
+        const size_t maxSteps = 200) :
+      maxSteps(maxSteps),
+      stepsPerformed(0)
+  { 
+    MazeGeneration(rows, columns);
+    
+    directions = std::unordered_map<Action::actions, arma::vec>({
+            { Action::actions::left,  arma::vec({0, -1}) },
+            { Action::actions::right,  arma::vec({0, 1}) },
+            { Action::actions::up,  arma::vec({-1, 0}) },
+            { Action::actions::down,  arma::vec({1, 0}) }
+        });
+
+  }
+
+  /**
+   * Generate Maze using Random Iterative DFS Algorithm with 
+   * 0.4 * rows * column steps
+   * @param rows number of rows in maze.
+   * @param columns number of columns in maze.
+   * 
+   */
+  void MazeGeneration(const size_t rows, const size_t columns)
+  {
+    maze = arma::mat(rows,columns);
+    arma::umat visited;
+    visited.zeros(rows, columns);
```

Comment:
This can be a one liner. 

`arma::umat visited(rows, columns, arma::fill::zeros)`